### ISSUE TITLE
More readily interpretable marginal effects with ordinal dependent variables

### ISSUE THREAD

**Issue by jackobailey** [paul-buerkner/brms#491]
I love BRMS and the marginal_effects functions. But I feel that it could be improved when it comes to ordinal dependent variables.

In particular, I think that the grid plots are difficult to interpret. Often, there might be meaningful differences in probability of selection that are easy to miss because the eye simply isn't built to differentiate such fine-grained differences in hue.

It seems to me that we have already solved the problem of how to plot ordinal outcomes -- item characteristic curves. Although here x-axis represents a latent score rather than an observed one. Rather than plotting grids, could you do something similar? The y-axis could be the probability of selecting a given response, the x-axis the independent variable we want marginal effects for, with a curve plotted for each possible ordered response. You could extend this to categorical or ordinal independent variables too. For instance, you could have probability on the y-axis, each predictor on the x-axis, and point estimates for each ordered response grouped by predictor rather than curves. This would also make it possible to include bands to characterise uncertainty that the current plots miss.

On a simpler note, as things currently stand the grid plots also seem to be plotting marginal effects that are out of sample. I have a predictor that ranges from .3 to .8. If I pretend the dependent variable is continuous, the marginal effects are within this range. If I treat it as ordinal, the marginal effects begin around 0 and go up to 1, despite those values not being included in the data.

**Comment by paul-buerkner** [comment]
Can you show me an example plot to illustrate what you have in mind?

**Comment by jackobailey** [comment]
I was thinking something like this, but with the independent variable that you want the effects for on the x-axis:

![category-characteristic-curve-for-item-b5-with-disordered-thresholds](https://user-images.githubusercontent.com/37049065/43836717-20f4a2f0-9b0e-11e8-99e0-007ae8561e28.png)

Let's imagine a toy example: income ~ age, and income is split into quartiles. In this example, we'd have age along the x-axis (as is currently the case), probability of selecting one of the 4 possible responses on the y-axis, and curves for each response.

So, basically, flip the placement of the ordered responses and probability in the current model so that the latter is on an axis and the former is in the legend.

**Comment by paul-buerkner** [comment]
I think this is doable. Let me think of a good way to integrate this in brms.

**Comment by jackobailey** [comment]
Brilliant! I imagine you could do something similar if you wanted effects for ordered/categorical predictors too, but instead of plotting curves you'd plot point estimates for each response.

**Comment by strengejacke** [comment]
I have implemented such plot-style in _ggeffects_, @paul-buerkner maybe you find some inspiration on how to transform the data or prepare the plot?

```r
library(MASS)
options(contrasts = c("contr.treatment", "contr.poly"))
housing$dummy <- rnorm(n = nrow(housing), mean = 30, sd = 10)
house.plr <- brm(Sat ~ Infl + Type + Cont + dummy, data = housing, family = cumulative)
```

```r
ggpredict(house.plr, c("dummy", "Type")) %>% plot()
```
![rp1](https://user-images.githubusercontent.com/26301769/43996817-7f339442-9dcb-11e8-952f-810b440899b0.png)

`ggpredict()` will also plot interaction-terms, which is currently not possible with `marginal_effects()` for `ordinal = TRUE`.

```r
house.plr <- brm(Sat ~ Infl * dummy + Type + Cont, data = housing, family = cumulative)
ggpredict(house.plr, c("dummy", "Infl")) %>% plot()
``` 

![rp2](https://user-images.githubusercontent.com/26301769/43996834-0d1de3ac-9dcc-11e8-9a22-904ac5e9ba8b.png)

(plot looks quite weird in this artificial example...)

**Comment by paul-buerkner** [comment]
Thanks I will take a look!

It is not quite correct, though, that we can't plot interactions with ordinal = TRUE, we just have to use the `conditions` argument to use faceting just as in the plot you show above.

**Comment by mvuorre** [comment]
This would be a very nice feature. You can do these quite easily with @mjskay 's tidybayes package (<https://mjskay.github.io/tidybayes/articles/tidy-brms.html#ordinal-models>). (Might look at the code over there for a potential implementation.)

**Comment by paul-buerkner** [comment]
Looks great! I am unsure about the interface though. Perhaps we need a new argument, let's call it `categorical`, taking on three possible values `"no"`, `"lines"` and `"surface"`  where the second generates the plots discusse here and the third generated the current surface kind of plots. However, I am honestly not sure if we should keep those `surface` ones altogether in the long run as they seem to be rather confusing for users. 

Any comments and suggestions are welcome.

**Comment by jackobailey** [comment]
I'd argue that you get rid of the `surface` plots altogether and switch the `ordinal == T` argument to `categorical == T` instead. This is important because, if you apply this to categorical outcomes generally and not just ordinal outcomes, for ordinal outcomes the output would be lines/curves and for categorical outcomes the output would be point estimates. Having a single argument that works for both seems parsimonious.

**Comment by paul-buerkner** [comment]
I agree, but I can't just get rid of `ordinal` immediately for reasons of backwards compatibility. Instead, I will deprecate it and likely remove it in brms 3.0

**Comment by paul-buerkner** [comment]
I am making good progress with the implementations. It's much easier than expected. A shame I haven't done this earlier...

With regard to handling the `ordinal` argument, I think we have two options.

1. Keep the functionality of `ordinal` as is but add a deprecation warning (and remove the argument in brms 3.0).
2. Make `ordinal` a deprecated alias of the new `categorical` argument. This will drastically simplify the code and will immediately advertise the new categorical plots discussed here. However, these ordinal surface plots will no longer be available thus breaking backwards compatibility. Usually, I want to avoid that but given that the new plots contain the same information and more (uncertainty estimates) I am tempted to do it.

@mvuorre What's you opinion on these two options?

**Comment by strengejacke** [comment]
IMHO, I would go for 2)

**Comment by mvuorre** [comment]
Even though the new figures are more informative I'd still go with 1, to not immediately break backward compatibility. In 3.0 you could then remove `ordinal` or make it an alias to `categorical`.

In any case I look forward to trying out the new figures.

**Comment by paul-buerkner** [comment]
The plots discussed in this issue are now implemented in the dev version of brms by means of the `categorical` argument in `marginal_effects`. Argument `ordinal` remains usable but is now deprecated.
